File: client/ios/DivKit/Actions/SetVariableActionHandler.swift
import Foundation

final class SetVariableActionHandler {
  func handle(_ action: DivActionSetVariable, context: DivActionHandlingContext) {
    let expressionResolver = context.expressionResolver
    guard let variableName = action.resolveVariableName(expressionResolver) else {
      return
    }

    guard let variableValue = action.value.resolveVariableValue(expressionResolver) else {
      return
    }

    context.variablesStorage.update(
      path: context.path,
      name: DivVariableName(rawValue: variableName),
      value: variableValue
    )
  }
}

File: client/android/compose/src/main/kotlin/com/yandex/div/compose/actions/SetVariableActionHandler.kt
package com.yandex.div.compose.actions

import com.yandex.div.compose.DivReporter
import com.yandex.div.data.VariableMutationException
import com.yandex.div.internal.actions.DivUntypedAction
import com.yandex.div.internal.variables.castAndSetValue
import com.yandex.div.internal.variables.evaluate
import com.yandex.div2.DivActionSetVariable
import javax.inject.Inject

internal class SetVariableActionHandler @Inject constructor(
    private val reporter: DivReporter
) {

    fun handle(
        context: DivActionHandlingContext,
        action: DivActionSetVariable
    ) {
        val expressionResolver = context.expressionResolver
        val variableName = action.variableName.evaluate(expressionResolver)
        val variable = expressionResolver.getVariable(variableName)
        if (variable == null) {
            reporter.reportError("Unknown variable: $variableName")
            return
        }

        try {
            variable.castAndSetValue(action.value.evaluate(expressionResolver))
        } catch (e: VariableMutationException) {
            reporter.reportError(e)
        }
    }

    fun handle(
        context: DivActionHandlingContext,
        action: DivUntypedAction.SetVariable
    ) {
        val variableName = action.name
        val variable = context.expressionResolver.getVariable(variableName)
        if (variable == null) {
            reporter.reportError("Unknown variable: $variableName")
            return
        }

        try {
            variable.set(action.value)
        } catch (e: VariableMutationException) {
            reporter.reportError(e)
        }
    }
}
